docs/src/main/sphinx/client/client-protocol.md

```diff
@@ -0,0 +1,145 @@
+# Client protocol
+
+The Trino client protocol is a HTTP-based protocol that allows
+[clients](/client) to submit SQL queries and receive results.
+
+The protocol is a sequence of REST API calls to the
+[coordinator](trino-concept-coordinator) of the Trino
+[cluster](trino-concept-cluster). Following is a high-level overview:
+
+1. Client submits SQL query text to the coordinator of the Trino cluster.
+2. The coordinator starts processing the query.
+3. The coordinator returns a result set and a URI `nextUri` on the coordinator.
+4. The client receives the result set and initiates another request for more
+   data from the URI `nextUri`.
+5. The coordinator continues processing the query and returns further data with
+   a new URI.
+6. The client and coordinator continue with steps 4. and 5. until all
+   result set data is returned to the client or the client stops requesting
+   more data.
+
+The client protocol supports two modes. Configure the [spooling
+protocol](protocol-spooling) for optimal throughput for your clients.
+
+(protocol-spooling)=
+## Spooling protocol
+
+The spooling protocol uses an object storage location to store the data for
+retrieval by the client. The coordinator and all workers can write result set
+data to the storage in parallel. The coordinator only provides the URLs to all
+the individual data segments on the object storage to the cluster. The spooling
+protocol also allows compression of the data.
+
+Data on the object storage is automatically removed after download by the
+client.
+
+The spooling protocol has the following characteristics, compared to the [direct
+protocol](protocol-direct).
+
+* Provides higher throughput for data transfer, specifically for queries that
+  return more data.
+* Results in faster query processing completion on the cluster, independent of
+  the client retrieving all data, since data is read from the object storage.
+* Requires object storage and configuration on the Trino cluster.
+* Reduces CPU and I/O load on the coordinator.
+* Automatically falls back to the direct protocol for queries that don't benefit
+  from using the spooling protocol.
+* Requires newer client drivers or client applications that support the spooling
+  protocol and actively request usage of the spooling protocol.
+* Clients must have access to the object storage.
```

Comment:
This mixes two things:
* The protocol does not require clients to have access to object storage. They need to have access to whatever endpoints are referenced by the URLs returned in the responses. These could be object storage, endpoints in the coordinator, workers, etc.
* Object storage is just an implementation strategy for realizing the protocol on the server side. When that implementation is used by the server, then clients need to have access to the object storage endpoints.

Reply:
okay .. that makes sense .. how do we write that though so its not completely abstract or super long. 

* Clients must have access to the spooled segments. Their locations are specified as 
  URLs to external object storage, the coordinator, or workers in the cluster.

Thats a lot of hedging.

Reply:
Also I can work that into https://github.com/trinodb/trino/pull/24288 or another PR .. once we figured out what we want to write.